### PR TITLE
drivers: add sama5d2 PIOBU driver

### PR DESCRIPTION
**Pull request by clementleger**
On sama5d2, the PIOBU (AKA secumod) module allows to configure some tampering inputs. This series adds support for the PIOBU module gpios and for the tampering detection module. In case of a tampering detection, the interrupt handler will panic.

### CODE REVIEW

**Review by jforissier — APPROVED** [review]
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/drivers/atmel_piobu.c [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+#define SECUMOD_MAX_PINS		8
+#define SECUMOD_PIN_MASK		(BIT(SECUMOD_MAX_PINS) - 1)
+#define SECUMOD_PIN_SHIFT		16
+#define SECUMOD_PIN_VAL(pin)		BIT(SECUMOD_PIN_SHIFT + (pin))
+
+#define DT_GPIO_CELLS			2
+
+#define SECUMOD_CR			0x0
+#define SECUMOD_CR_KEY_SHIFT		16
+#define SECUMOD_CR_KEY			(0x89CA << SECUMOD_CR_KEY_SHIFT)
```
Comment (jenswi-linaro) [inline core/drivers/atmel_piobu.c:28]: Prefer using `SHIFT_U32()`. The result here seems to be implementation defined.

File: core/drivers/atmel_piobu.c [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t sr = io_read32(secumod_base + SECUMOD_SR);
+
+	for (i = 0; i < SECUMOD_MAX_PINS; i++) {
+		if (sr & SECUMOD_PIN_VAL(i))
+			EMSG("Detected tampering on pin %d", i);
+	}
+
+	if (sr & SECUMOD_SR_JTAG)
+		EMSG("JTAG tampering attempt\n");
```
Comment (jenswi-linaro) [inline core/drivers/atmel_piobu.c:216]: `\n` not needed. Please fix all the occurences in this PR.

**Review by jenswi-linaro — COMMENTED** [review]
The the comments addressed, please apply:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by clementleger** [comment]
Version 3:
- Fix some comments
- Fix struct tm initialization

File: core/drivers/atmel_piobu.c [hunk 3]
```diff
@@ -0,0 +1,383 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2022 Microchip
+ *
+ * Driver for AT91 PIOBU
+ */
+
+#include <assert.h>
+#include <dt-bindings/gpio/atmel,piobu.h>
+#include <drivers/atmel_rtc.h>
```
Comment (etienne-lms) [inline core/drivers/atmel_piobu.c:10]: swap the 2 above lines

File: core/include/dt-bindings/gpio/atmel,piobu.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
+/* SPDX-License-Identifier: (GPL-2.0 OR BSD-2-Clause) */
+/*
+ * Copyright (C) 2021 Microchip
+ */
+
+#ifndef _DT_BINDINGS_ATMEL_PIOBU_H
```
Comment (etienne-lms) [inline core/include/dt-bindings/gpio/atmel,piobu.h:6]: _DT_BINDINGS_GPIO_ATMEL_PIOBU_H ?

File: core/include/dt-bindings/gpio/atmel,piobu.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
+/* SPDX-License-Identifier: (GPL-2.0 OR BSD-2-Clause) */
+/*
+ * Copyright (C) 2021 Microchip
```
Comment (etienne-lms) [inline core/include/dt-bindings/gpio/atmel,piobu.h:3]: 2022

File: core/include/dt-bindings/gpio/atmel,piobu.h (hunk 4 above)
Comment (clementleger) [reply core/include/dt-bindings/gpio/atmel,piobu.h:6]: Ok

**Comment by clementleger** [comment]
Ping ?

**Comment by jforissier** [comment]
@etienne-lms any comments?

**Comment by clementleger** [comment]
Ping ?

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`
(sorry for the delay)

**Pull request merged**